File: java/src/main/java/org/tvarley/euler/solutions/Solution087.java
// https://projecteuler.net/problem=87
// Prime Power Triples
//
// Find the number of numbers below fifty million that can be expressed as the sum of a prime square, prime cube, and prime fourth power.
//
// Answer: 1097343
//
// Completed on: 2024-12-04
// Time: ~100ms
// Optimizations: Sieve for primes, nested loops with early breaks
// Notes: Used HashSet to store unique sums.

package org.tvarley.euler.solutions;

import org.tvarley.euler.Solution;
import java.util.HashSet;
import java.util.Set;

public class Solution087 implements Solution {
    private boolean[] isPrime;
    private int maxPrime;

    public String solve() {
        final long LIMIT = 50000000;
        maxPrime = 0;
        while ((long) maxPrime * maxPrime < LIMIT) maxPrime++;
        maxPrime--;
        generatePrimes(maxPrime);
        Set<Long> numbers = new HashSet<>();
        for (int p4 = 2; isPrime[p4]; ++p4) {
            long fourth = (long) p4 * p4 * p4 * p4;
            if (fourth >= LIMIT) break;
            for (int p3 = 2; isPrime[p3]; ++p3) {
                long cube = (long) p3 * p3 * p3;
                if (fourth + cube >= LIMIT) break;
                for (int p2 = 2; isPrime[p2]; ++p2) {
                    long square = (long) p2 * p2;
                    long sum = fourth + cube + square;
                    if (sum >= LIMIT) break;
                    numbers.add(sum);
                }
            }
        }
        return Long.toString(numbers.size());
    }

    private void generatePrimes(int maxN) {
        isPrime = new boolean[maxN + 1];
        for (int i = 2; i <= maxN; ++i) {
            isPrime[i] = true;
        }
        for (long i = 2; i <= maxN; ++i) {
            if (isPrime[(int) i]) {
                for (long j = i * i; j <= maxN; j += i) {
                    isPrime[(int) j] = false;
                }
            }
        }
    }
}

File: javascript/src/euler/solution087.js
// https://projecteuler.net/problem=87
// Prime power triples
//
// The smallest number expressible as the sum of a prime square, prime cube, and prime fourth power is 28. In fact, there are exactly four numbers below fifty that can be expressed in such a way:
//
// 28 = 2² + 2³ + 2⁴
// 33 = 3² + 2³ + 2⁴
// 49 = 5² + 2³ + 2⁴
// 47 = 2² + 3³ + 2⁴
//
// How many numbers below fifty million can be expressed as the sum of a prime square, prime cube, and prime fourth power?
//
// The final answer to the problem is 1097343.
//
// Execution time measurements: Approximately 200 ms.
//
// Details on any optimizations made during development: Sieve for prime generation, nested loops with early breaks, Set for uniqueness.
// Interesting notes or insights from the AI implementation process: Adapted sieve and nested loops from C++, using BigInt for power calculations to avoid overflow.

function prime_power_triples() {
  const LIMIT = 50000000n;
  let max_prime = 0;
  while (BigInt(max_prime) * BigInt(max_prime) < LIMIT) max_prime++;
  max_prime--;
  const is_prime = new Array(max_prime + 1).fill(true);
  is_prime[0] = is_prime[1] = false;
  const primes = [];
  for (let i = 2; i <= max_prime; i++) {
    if (is_prime[i]) {
      primes.push(i);
      for (let j = i * i; j <= max_prime; j += i) {
        is_prime[j] = false;
      }
    }
  }
  const numbers = new Set();
  for (const p4 of primes) {
    const fourth = BigInt(p4) ** 4n;
    if (fourth >= LIMIT) break;
    for (const p3 of primes) {
      const cube = BigInt(p3) ** 3n;
      if (fourth + cube >= LIMIT) break;
      for (const p2 of primes) {
        const square = BigInt(p2) ** 2n;
        const sum = fourth + cube + square;
        if (sum >= LIMIT) break;
        numbers.add(sum);
      }
    }
  }
  return numbers.size;
}

module.exports = {
  answer: () => prime_power_triples()
};